### PR TITLE
nvidia_x11.dc_535: 535.129.03

### PR DESCRIPTION
**Pull request by timothyklim**
## Description of changes

https://docs.nvidia.com/datacenter/tesla/tesla-release-notes-535-129-03/index.html

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by timothyklim** [comment]
cc @jonringer @kiskae @edwtjo

**Comment by Kiskae** [comment]
Any specific reason to use this version and not https://docs.nvidia.com/datacenter/tesla/tesla-release-notes-535-129-03/index.html ?

**Comment by timothyklim** [comment]
> Any specific reason to use this version and not https://docs.nvidia.com/datacenter/tesla/tesla-release-notes-535-129-03/index.html ?

Yes, I'm planning to create another PR with CUDA 12.2.2: https://github.com/timothyklim/nixpkgs/commit/444fd6f94aeca98e8b54eaff49b75f0fd81c4795 where both versions the same or I miss something?

**Comment by timothyklim** [comment]
> > Any specific reason to use this version and not https://docs.nvidia.com/datacenter/tesla/tesla-release-notes-535-129-03/index.html ?
> 
> Yes, I'm planning to create another PR with CUDA 12.2.2: [timothyklim@444fd6f](https://github.com/timothyklim/nixpkgs/commit/444fd6f94aeca98e8b54eaff49b75f0fd81c4795) where both versions the same or I miss something?

Also, new PR will close issue https://github.com/NixOS/nixpkgs/issues/259887 . I tested it on server with 8xA100 GPUs, everything works smoothly.

**Comment by Kiskae** [comment]
> > Any specific reason to use this version and not https://docs.nvidia.com/datacenter/tesla/tesla-release-notes-535-129-03/index.html ?
> 
> Yes, I'm planning to create another PR with CUDA 12.2.2: https://github.com/timothyklim/nixpkgs/commit/444fd6f94aeca98e8b54eaff49b75f0fd81c4795 where both versions the same or I miss something?

The cuda version support is the same on both drivers, but the one I linked is a more recent release.

**Comment by timothyklim** [comment]
> [535-129-03/index.html](https://docs.nvidia.com/datacenter/tesla/tesla-release-notes-535-129-03/index.html) ?

But there no fabricmanager for that version: https://developer.download.nvidia.com/compute/cuda/redist/fabricmanager/linux-x86_64/

**Comment by Kiskae** [comment]
It appears there is a different repository of releases here: https://developer.download.nvidia.com/compute/nvidia-driver/redist/fabricmanager/linux-x86_64/

I'm not sure where this is documented though.

**Comment by timothyklim** [comment]
> It appears there is a different repository of releases here: https://developer.download.nvidia.com/compute/nvidia-driver/redist/fabricmanager/linux-x86_64/
> 
> I'm not sure where this is documented though.

Updated with latest version

File: pkgs/os-specific/linux/nvidia-x11/generic.nix [hunk 1]
```diff
@@ -53,7 +53,7 @@
 , firmware ? openSha256 != null || useFabricmanager
   # Whether the user accepts the NVIDIA Software License
 , config
-, acceptLicense ? config.nvidia.acceptLicense or false
+, acceptLicense ? config.nvidia.acceptLicense or true
```
Comment (Kiskae) [inline pkgs/os-specific/linux/nvidia-x11/generic.nix:56]: While this makes testing easier, I assume this is not intended to be committed?

File: nixos/modules/hardware/video/nvidia.nix [hunk 2]
```diff
@@ -584,24 +586,50 @@ in {
...
+                wantedBy = [ "multi-user.target" ];
+                unitConfig.After = [ "network-online.target" ];
+                unitConfig.Requires = [ "network-online.target" ];
+                serviceConfig = {
+                  Type = "forking";
+                  TimeoutStartSec = 240;
+                  ExecStart = let
+                    nv-fab-conf = settingsFormat.generate "fabricmanager.conf" cfg.datacenter.settings;
+                    in
+                      nvidia_x11.fabricmanager + "/bin/nv-fabricmanager -c " + nv-fab-conf;
```
Comment (Kiskae) [inline nixos/modules/hardware/video/nvidia.nix:609]:
```suggestion
"${lib.getExe nvidia_x11.fabricmanager} -c ${nv-fab-conf}";
```

**Review by Kiskae — COMMENTED** [review]
Some minor cosmetic changes suggested, but functionally looks good.

File: pkgs/os-specific/linux/nvidia-x11/generic.nix (hunk 1 above)
Comment (jonringer) [reply pkgs/os-specific/linux/nvidia-x11/generic.nix:56]: Agreed, this shouldn't be upstream to nixpkgs. The point is to ensure that people are "opting into" nvidia software license.

Comment (timothyklim) [reply pkgs/os-specific/linux/nvidia-x11/generic.nix:56]: I forget to delete it after tests in repl.

**Comment by timothyklim** [comment]
Can it be backported to 23.11?

**Comment by timothyklim** [comment]
> Some minor cosmetic changes suggested, but functionally looks good.

I updated commit

File: nixos/modules/hardware/video/nvidia.nix [hunk 3]
```diff
@@ -70,6 +71,7 @@ in {
           NVSWITCH_FAILURE_MODE=0;
           ABORT_CUDA_JOBS_ON_FM_EXIT=1;
           TOPOLOGY_FILE_PATH=nvidia_x11.fabricmanager + "/share/nvidia-fabricmanager/nvidia/nvswitch";
+          DATABASE_PATH=nvidia_x11.fabricmanager + "/share/nvidia-fabricmanager/nvidia/nvswitch";
```
Comment (Kiskae) [inline nixos/modules/hardware/video/nvidia.nix:74]: Example should be updated to match actual configuration

**Review by Kiskae — APPROVED** [review]

File: nixos/modules/hardware/video/nvidia.nix (hunk 3 above)
Comment (timothyklim) [reply nixos/modules/hardware/video/nvidia.nix:74]: > Example should be updated to match actual configuration

I cannot compile then, I tried before.

Comment (Kiskae) [reply nixos/modules/hardware/video/nvidia.nix:74]: I think you just need to escape the formatting symbol like `''$`

**Review by Kiskae — APPROVED** [review]

**Comment by timothyklim** [comment]
gentle ping @Kiskae @edwtjo

**Review by NickCao — APPROVED** [review]

**Comment by Kiskae** [comment]
Result of `nixpkgs-review pr 273876 --extra-nixpkgs-config '{ nvidia.acceptLicense = true; }'` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>

**Comment by NickCao** [comment]
> Result of `nixpkgs-review pr 273876 --extra-nixpkgs-config '{ nvidia.acceptLicense = true; }'` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
> 2 packages blacklisted:

Um? Shouldn't there be new packages.

**Comment by Kiskae** [comment]
Oh, it looks like this only adds an option to `linuxPackages.nvidiaPackages`, which isn't evaluated by default.

**Comment by Kiskae** [comment]
`nixpkgs-review pr 273876 --extra-nixpkgs-config '{ nvidia.acceptLicense = true; }' --eval local` for future reference.

**Pull request merged**